### PR TITLE
Automatically derive serde impls

### PR DESCRIPTION
**Pull request by cpcloud**
cc @nagisa

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -8,14 +8,15 @@ license-file = "LICENSE"
...
+keywords = ["tensorflow", "protobuf", "config", "serde"]
 readme = "README.md"
 
 [dependencies]
 prost = "0.6.1"
-prost-types = "0.6.1"
 thiserror = "1.0.20"
-serde = { version = "1", features = ["derive"], optional = true }
+serde = { version = "1", features = ["derive"] }
+tensorflow_proto_derive = { path = "./tensorflow_proto_derive" }
```
Comment (nagisa) [inline Cargo.toml]: Specify a version as well.

File: tensorflow_proto_derive/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    let name = ast.ident.clone();
+
+    match syn::Item::from(ast) {
+        syn::Item::Struct(_) => {
+            quote::quote! {
+                paste::paste! {
+                    #[derive(Default, Clone, serde::Deserialize, serde::Serialize)]
+                    #[serde(transparent, default)]
+                    struct [<Serde #name>] {
+                        inner: #name
```
Comment (nagisa) [inline tensorflow_proto_derive/src/lib.rs]: Probably need a test to ensure this is not mutually recursive.

Comment (nagisa) [reply tensorflow_proto_derive/src/lib.rs]: (I suspect that `serde(transparent)` might be calling `#name::deserialize(..)` which then calls `Serde#name::deserialize(..)` which then calls `#name::deserialize(..)`.

Comment (cpcloud) [reply tensorflow_proto_derive/src/lib.rs]: Hm, now I'm not entirely sure how to delegate here.

Comment (cpcloud) [reply tensorflow_proto_derive/src/lib.rs]: IOW, how do I tell `serde` to use the `inner` serde?

Comment (nagisa) [reply tensorflow_proto_derive/src/lib.rs]: I think you might want to replicate the fields of the original type within `Serde#name` type and then convert from one to another in the `deserialize` method of `#name`.

Comment (cpcloud) [reply tensorflow_proto_derive/src/lib.rs]: Ultimately ended up depending on upstream `prost` to pickup this commit: https://github.com/danburkert/prost/commit/d9547f173246480e69120300f41c190257e0b63f so I could implement this whole thing as an attribute macro, greatly simplifying the implementation.

**Comment by cpcloud** [comment]
@nagisa This could use another pass, at your leisure.

File: src/lib.rs [hunk 3]
```diff
@@ -15,24 +14,40 @@ pub enum Error {
...
     msg.encode(&mut bytes).map_err(Error::Encode)?;
     Ok(bytes)
 }
+
+#[cfg(test)]
+mod tests {
+    use super::{into_bytes, tensorflow};
+
+    #[test]
+    fn test_serde() {
```
Comment (nagisa) [inline src/lib.rs]: Add also a test that deserializes from something like

```json
{  "gpu_options": { "allow_growth": true } }
```

**Review by nagisa — APPROVED** [review]

**Pull request merged**